### DISCUSSION TITLE
WHY MY FOREIGN KEY VALUES ALWAYS COMES NULL?

### DISCUSSION THREAD

**Discussion by justnashy** [community/community#133190] [Programming Help]
### Select Topic Area

Question

### Body

HELLO ALL,

I AM STUCK WITH UNIVERSITY ASSESSMENT AND NEW TO MYSQL.

WHEN I TRY TO DISPLAY TABLE DATA MY FOREIGN KEY VALUES ALWAYS COMES AS NULL,
I HAVE NOT BEEN DELETED ANY OF PARENT TABLE AND HERE ARE THE CODES,
COULD SOMEONE ASSIST ME WITH FIX THIS UP FOR ME PLEASE?

CHILD TABLE:

CREATE TABLE PLAYERS (
FIRST_NAME VARCHAR(30)  NOT NULL,
LAST_NAME VARCHAR(30) NOT NULL,
GENDER ENUM('M','F') NOT NULL,
PLAYER_ID INT, FOREIGN KEY (PLAYER_ID) REFERENCES PLAYERS_NEW(PLAYER_ID)
);

PARENT TABLE:

CREATE TABLE PLAYERS_NEW (
PLAYER_ID INT PRIMARY KEY AUTO_INCREMENT NOT NULL,
FIRST_NAME_X VARCHAR(30) NOT NULL,
LAST_NAME_X VARCHAR(30) NOT NULL,
GENDER_X ENUM ('M','F') NOT NULL
);

<img width="1440" alt="Screenshot 2024-07-23 at 3 27 56 PM" src="https://github.com/user-attachments/assets/4da5fcee-2e64-4bd1-90dd-6297de3c0217">

CHILD TABLE DATA;
INSERT INTO PLAYERS (FIRST_NAME, LAST_NAME, GENDER)
VALUES  
("Adelina","Dragonscream","F"),
 ("Adelina","Dragonscream","F"),
  ("Adelina","Dragonscream","F"),
  ("Adelina","Dragonscream","F"),
     ("Agaell","Clavira","F"),
         ("Alicia","Stormmane","F"),
         ("Alicia","Stormmane","F"),
         ("Cara","the Menace","F"),
         ("Cara","the Menace","F"),
         ("Cara","the Menace","F"),
         ("Cara","the Menace","F"),
         ("Cara","the Menace","F"),
         ("Cara","the Menace","F"),
         ("Cara","the Menace","F"),
         ("Cara","the Menace","F"),
		 ("Davis","Smokes","M"),
         ("Davis","Smokes","M"),
         ("Davis","Smokes","M"),
         ("Pikipius","Tepewrick","M"),
         ("Pikipius","Tepewrick","M"),
         ("Pikipius","Tepewrick","M"),
         ("Pikipius","Tepewrick","M"),
		 ("Pikipius","Tepewrick","M"),
         ("Pikipius","Tepewrick","M"),
         ("Pikipius","Tepewrick","M");

CHILD TABLE DATA:
INSERT INTO PLAYERS_NEW(FIRST_NAME_X, LAST_NAME_X, GENDER_X)
VALUES   ("ADELINA","DRAGONSCREAM","F"),
	     ("AGAELL","CLAVIRA","F"),
	     ("ALICIA","STORMMANE","F"),
         ("CARA","THE MENACE","F"),
         ("DAVIS","SMOKES","M"),
         ("PIKIPIUS","TEPEWRICK","M");

**Comment by Damandeep-S** [comment]
The issue with the PLAYERS table having PLAYER_ID values as NULL is due to how you're inserting data. You need to ensure that the PLAYER_ID in the PLAYERS table corresponds to an existing PLAYER_ID in the PLAYERS_NEW table. The PLAYER_ID in the PLAYERS table should be a reference to the PLAYER_ID in the PLAYERS_NEW table.

**Comment by Damandeep-S** [comment]
try this 

`
CREATE TABLE PLAYERS_NEW (
    PLAYER_ID INT PRIMARY KEY AUTO_INCREMENT NOT NULL,
    FIRST_NAME_X VARCHAR(30) NOT NULL,
    LAST_NAME_X VARCHAR(30) NOT NULL,
    GENDER_X ENUM ('M','F') NOT NULL
);


CREATE TABLE PLAYERS (
    PLAYER_ID INT, 
    FIRST_NAME VARCHAR(30) NOT NULL,
    LAST_NAME VARCHAR(30) NOT NULL,
    GENDER ENUM('M','F') NOT NULL,
    FOREIGN KEY (PLAYER_ID) REFERENCES PLAYERS_NEW(PLAYER_ID)
);


INSERT INTO PLAYERS_NEW (FIRST_NAME_X, LAST_NAME_X, GENDER_X)
VALUES 
("ADELINA", "DRAGONSCREAM", "F"),
("AGAELL", "CLAVIRA", "F"),
("ALICIA", "STORMMANE", "F"),
("CARA", "THE MENACE", "F"),
("DAVIS", "SMOKES", "M"),
("PIKIPIUS", "TEPEWRICK", "M");


INSERT INTO PLAYERS (PLAYER_ID, FIRST_NAME, LAST_NAME, GENDER)
VALUES
((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ADELINA' AND LAST_NAME_X = 'DRAGONSCREAM' AND GENDER_X = 'F'), "Adelina", "Dragonscream", "F"),
((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'AGAELL' AND LAST_NAME_X = 'CLAVIRA' AND GENDER_X = 'F'), "Agaell", "Clavira", "F"),
((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ALICIA' AND LAST_NAME_X = 'STORMMANE' AND GENDER_X = 'F'), "Alicia", "Stormmane", "F"),
((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'), "Cara", "the Menace", "F"),
((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'DAVIS' AND LAST_NAME_X = 'SMOKES' AND GENDER_X = 'M'), "Davis", "Smokes", "M"),
((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'), "Pikipius", "Tepewrick", "M");
`

**Reply by justnashy** [reply]
HELLO DAMANDEEP,
THANK YOU FOR YOUR HELP AND IT WORKED 
I HAVE PARTICULAR EXCEL DATASHEET WHICH PROVIDED FROM THE UNIVERSITY,
I WILL ATTCH THAT HERE, 
I AM NOT ALLOWED TO IMPORT SPREADSHEET INTO MYSQL. ONLY INSERT DATA MANUAL WAY.

I HAVE TRIED TO NORMALIZE PLAYER DATA . USING TWO TABLES

[3c30fa2f-3d5f-496c-baae-266c06abdd15.xlsx](https://github.com/user-attachments/files/16344262/3c30fa2f-3d5f-496c-baae-266c06abdd15.xlsx)

**Reply by justnashy** [reply]
I TRIED THIS WAY BUT ERROR OCCURRED,

Error Code: 1242. Subquery returns more than 1 row


INSERT INTO PLAYERS (PLAYER_ID,FIRST_NAME, LAST_NAME, GENDER)
VALUES   ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ADELINA' AND LAST_NAME_X = 'DRAGONSCREAM' AND GENDER_X = 'F'),"Adelina","Dragonscream","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ADELINA' AND LAST_NAME_X = 'DRAGONSCREAM' AND GENDER_X = 'F'),"Adelina","Dragonscream","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ADELINA' AND LAST_NAME_X = 'DRAGONSCREAM' AND GENDER_X = 'F'),"Adelina","Dragonscream","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ADELINA' AND LAST_NAME_X = 'DRAGONSCREAM' AND GENDER_X = 'F'),"Adelina","Dragonscream","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'AGAELL' AND LAST_NAME_X = 'CLAVIRA' AND GENDER_X = 'F'),"Agaell","Clavira","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ALICIA' AND LAST_NAME_X = 'STORMMANE' AND GENDER_X = 'F'),"Alicia","Stormmane","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'ALICIA' AND LAST_NAME_X = 'STORMMANE' AND GENDER_X = 'F'),"Alicia","Stormmane","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'CARA' AND LAST_NAME_X = 'THE MENACE' AND GENDER_X = 'F'),"Cara","the Menace","F"),
		 ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'DAVIS' AND LAST_NAME_X = 'SMOKES' AND GENDER_X = 'M'), "Davis","Smokes","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'DAVIS' AND LAST_NAME_X = 'SMOKES' AND GENDER_X = 'M'), "Davis","Smokes","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'DAVIS' AND LAST_NAME_X = 'SMOKES' AND GENDER_X = 'M'), "Davis","Smokes","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'),"Pikipius","Tepewrick","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'),"Pikipius","Tepewrick","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'),"Pikipius","Tepewrick","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'),"Pikipius","Tepewrick","M"),
		 ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'),"Pikipius","Tepewrick","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'),"Pikipius","Tepewrick","M"),
         ((SELECT PLAYER_ID FROM PLAYERS_NEW WHERE FIRST_NAME_X = 'PIKIPIUS' AND LAST_NAME_X = 'TEPEWRICK' AND GENDER_X = 'M'),"Pikipius","Tepewrick","M");

**Reply by Damandeep-S** [reply]
The error occurs because the subquery is returning multiple rows, which is not allowed in this context. To resolve this, you need to ensure that each subquery returns a single row.

**Reply by justnashy** [reply]
yeah make sense, do we have any other option to take foreign key data from parent table PK?

**Reply by justnashy** [reply]
or do you have any idea about normalising that data in excel sheet? would happy to get advices or an idea if you have any thanks

**Reply by Damandeep-S** [reply]
Yes, there are other ways to handle foreign key data from the parent table's primary key in SQL. Additionally, normalizing data in an Excel sheet before importing it into a database can help in simplify the process

you can use a Temporary Table to Handle Foreign Keys or you can normalize data in excel

**Reply by Damandeep-S** [reply]
To normalize data , you have  parent Table (PLAYERS_NEW) Sheet and Child Table (PLAYERS) sheet then use MySQL Workbench or any other database tool to import the Excel sheets into MySQL tables.

**Comment by justnashy** [comment]
Yeah, thanks mate, i will try to do that and let ya know when i done that process

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by Damandeep-S** [reply]
can you marked it as answered or tell me if you need any more help

**Reply by justnashy** [reply]
hello Damandeep,
when I tried to import my excel sheet there is a error comes like this, I have no idea how to fix it as I am using macOS 

Error Code: 1290. The MySQL server is running with the --secure-file-priv option so it cannot execute this statement

**Comment by Malinhara** [comment]
To mitigate these issues:

Ensure Data Integrity: Validate that all required foreign key relationships are in place before performing inserts.
Set Constraints Correctly: Define your foreign key constraints with the appropriate ON DELETE and ON UPDATE actions (e.g., CASCADE, SET NULL) to handle situations where the referenced records are deleted or updated.
Order of Operations: Insert data in a sequence where primary keys are inserted first, followed by records containing foreign keys.

**Comment by Malinhara** [comment]
CREATE TABLE departments (
    id INT PRIMARY KEY,
    name VARCHAR(50) NOT NULL
);

CREATE TABLE employees (
    id INT PRIMARY KEY,
    name VARCHAR(50) NOT NULL,
    department_id INT,
    FOREIGN KEY (department_id) REFERENCES departments(id)
);



INSERT INTO departments (id, name) VALUES (1, 'HR');
INSERT INTO employees (id, name, department_id) VALUES (1, 'John Doe', NULL);
